File: java/src/main/java/it/giannibombelli/wsc2026/giftcard/application/usecases/GiftCardCrediting.java
package it.giannibombelli.wsc2026.giftcard.application.usecases;

import it.giannibombelli.wsc2026.common.utils.Require;

import it.giannibombelli.wsc2026.common.application.UseCase;
import it.giannibombelli.wsc2026.giftcard.application.commands.CreditGiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.events.GiftCardCredited;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCard;
import it.giannibombelli.wsc2026.giftcard.domain.giftcard.GiftCardId;
import it.giannibombelli.wsc2026.giftcard.domain.ports.GiftCardRepository;

import static java.util.Objects.requireNonNull;

public class GiftCardCrediting implements UseCase<CreditGiftCard, GiftCardCredited> {
    private final GiftCardRepository giftCardRepository;

    public GiftCardCrediting(GiftCardRepository giftCardRepository) {
        this.giftCardRepository = requireNonNull(giftCardRepository);
    }

    @Override
    public GiftCardCredited invoke(CreditGiftCard cmd) {
        Require.requireArgument(cmd, "command");
        GiftCardId giftCardId = cmd.aggregateId();

        GiftCard giftCard = giftCardRepository.findById(giftCardId)
            .orElseThrow(() -> new IllegalStateException("GiftCard not found for credit: " + giftCardId));

        GiftCardCredited credited = giftCard.credit(cmd.amount());

        giftCardRepository.save(giftCard);
        return credited;
    }
}

File: typescript/src/giftcard/application/usecases/giftCardCrediting.ts
// Caso d'uso per l'accredito di una gift card.

import type { UseCase } from '@/common/application/usecase.js';
import type { GiftCardCredited } from '../../domain/events/giftCardCredited.js';
import { GiftCardId } from '../../domain/giftcard/giftCardId.js';
import type { GiftCardRepository } from '../../domain/ports/giftCardRepository.js';
import { CreditGiftCard } from '../commands/creditGiftCard.js';
import { requireArgument } from '@/common/utils/requireArgument.js';
import { requireDependency } from '@/common/utils/requireDependency.js';

export class GiftCardCrediting implements UseCase<CreditGiftCard, GiftCardCredited> {
  private readonly _giftCardRepository: GiftCardRepository;

  constructor(giftCardRepository: GiftCardRepository) {
    requireDependency(giftCardRepository, "giftCardRepository");
    this._giftCardRepository = giftCardRepository;
  }

  invoke(command: CreditGiftCard): GiftCardCredited {
    requireArgument(command, 'command');

    const giftCardId: GiftCardId = command.aggregateId;
    const giftCard = this._giftCardRepository.findById(giftCardId);
    if (giftCard === null) {
      throw new Error(`GiftCard not found for credit: ${giftCardId.value}`);
    }

    const credited = giftCard.credit(command.amount);

    this._giftCardRepository.save(giftCard);
    return credited;
  }
}
